File: api/src/test/java/net/blueshell/api/factory/dto/survey/SurveyDTOFactory.java
package net.blueshell.api.factory.dto.survey;

import lombok.RequiredArgsConstructor;
import net.blueshell.api.common.enums.QuestionType;
import net.blueshell.api.dto.survey.QuestionDTO;
import net.blueshell.api.dto.survey.SurveyDTO;
import net.blueshell.api.factory.dto.BaseDtoFactory;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Factory for SurveyDTO test instances.
 */
@Component
@RequiredArgsConstructor
public class SurveyDTOFactory extends BaseDtoFactory<SurveyDTO> {

    private final QuestionDTOFactory questionFactory;

    @Override
    public Class<SurveyDTO> targetType() {
        return SurveyDTO.class;
    }

    @Override
    public SurveyDTO createBasic() {
        return createWithQuestionTypes(QuestionType.OPEN, QuestionType.RADIO);
    }

    public SurveyDTO createWithQuestionTypes(QuestionType... questionTypes) {
        SurveyDTO dto = new SurveyDTO();

        List<QuestionDTO> questions = Arrays.stream(questionTypes)
                .map(questionFactory::createByType)
                .collect(Collectors.toList());

        for (int i = 0; i < questions.size(); i++) {
            questions.get(i).setIdx((long) (i + 1));
        }

        dto.setQuestions(questions);
        dto.setResponseCount(0L);
        return dto;
    }

    public SurveyDTO createWithOpenQuestions(int count) {
        QuestionType[] types = new QuestionType[count];
        Arrays.fill(types, QuestionType.OPEN);
        return createWithQuestionTypes(types);
    }

    public SurveyDTO createWithMixedQuestions() {
        return createWithQuestionTypes(
                QuestionType.DESCRIPTION, QuestionType.RADIO, QuestionType.CHECKBOX, QuestionType.OPEN
        );
    }

    public SurveyDTO createWithMultipleChoiceOnly() {
        return createWithQuestionTypes(
                QuestionType.RADIO, QuestionType.RADIO, QuestionType.CHECKBOX
        );
    }
}

File: api/src/test/kotlin/net/blueshell/api/factory/dto/survey/SurveyDTOFactory.kt
package net.blueshell.api.factory.dto.survey

import net.blueshell.api.common.enums.QuestionType
import net.blueshell.api.dto.survey.QuestionDTO
import net.blueshell.api.dto.survey.SurveyDTO
import net.blueshell.api.factory.dto.BaseDtoFactory
import org.springframework.stereotype.Component

/**
 * Factory for SurveyDTO test instances.
 */
@Component
class SurveyDTOFactory(
    private val questionFactory: QuestionDTOFactory
) : BaseDtoFactory<SurveyDTO>() {

    override fun targetType(): Class<SurveyDTO> = SurveyDTO::class.java

    override fun createBasic(): SurveyDTO = createWithQuestionTypes(QuestionType.OPEN, QuestionType.RADIO)

    fun createWithQuestionTypes(vararg questionTypes: QuestionType): SurveyDTO {
        val dto = SurveyDTO()

        val questions = questionTypes.map { questionFactory.createByType(it) }.toMutableList()
        for (i in questions.indices) {
            questions[i].idx = (i + 1).toLong()
        }

        dto.questions = questions
        dto.responseCount = 0L
        return dto
    }

    fun createWithOpenQuestions(count: Int): SurveyDTO {
        val types = Array(count) { QuestionType.OPEN }
        return createWithQuestionTypes(*types)
    }

    fun createWithMixedQuestions(): SurveyDTO {
        return createWithQuestionTypes(
            QuestionType.DESCRIPTION, QuestionType.RADIO, QuestionType.CHECKBOX, QuestionType.OPEN
        )
    }

    fun createWithMultipleChoiceOnly(): SurveyDTO {
        return createWithQuestionTypes(
            QuestionType.RADIO, QuestionType.RADIO, QuestionType.CHECKBOX
        )
    }
}
